### ISSUE TITLE
[interop] Add support for Date/Time/Duration/TimeZone/... interop types

### ISSUE THREAD

**Issue by fniephaus** [oracle/truffleruby#2166]
TruffleRuby lacks support for some interop types, for example `Date`:

```smalltalk
Interop isDate: (Polyglot eval: 'ruby' string: 'DateTime.now.to_date') "returns `false`"
```

Some of these types may be implemented in Ruby, so adding interop support could be a bit tricky. In TruffleSqueak, we currently plan to dispatch interop messages directly on the object, so for example `Date` objects will respond to `#isDate` and `#asDate:`.

**Comment by eregon** [comment]
It should be fairly easy for `Time`, which is in core (and considered faster).
Date/DateTime is a stdlib: https://github.com/oracle/truffleruby/blob/master/lib/truffle/date.rb so we'd need to add `polyglot_date?`/`polyglot_time?` and `polyglot_as_date`/`polyglot_as_time` and implement it there + somehow convert from the very different internal representations.
There are actually quite a lot of date-related messages, notably also `isInstant` and `isTimeZone` stuff.
`Time` is using `java.time.ZonedDateTime` so hopefully it has all of that easily accessible.

I don't think Ruby has any object representing `isDuration`. Integer/Float are used instead for durations.

So I'd suggest to try to implement date-related messages for Time first, before trying it for date.rb which seems complicated.

**Comment by fniephaus** [comment]
Sounds good!

Why prefix interop messages with `polyglot_` rather than `interop_`? If someone looks at `interop_time?`, for example, it might be clearer that it's an interop message.

**Comment by eregon** [comment]
In TruffleRuby, we consider anything named "polyglot" to be public API, and anything named "interop" to be more internal APIs.
Anyway, such methods are not called by the user, but they can be defined by the user, like https://github.com/oracle/truffleruby/blob/5fd2b6b4872b2464425cda26eacda4f24b89fdca/spec/truffle/interop/fixtures/classes.rb#L110

I'll add support for date/time messages for Ruby's Time.

For Date/DateTime, one can easily workaround by using `Date#to_time` and `DateTime#to_time`.

**Comment by eregon** [comment]
Implemented for `Time` in 159cee8c8b93aec896a62027e300108226793117.
